### PR TITLE
LNURL Withdraw

### PR DESCRIPTION
**Pull request by michaelWuensch**
## Description

This PR adds the LNRUL-withdraw flow.
LNURL-pay is not part of this PR.

Please note:
Reading a LNURL-withdraw address via NFC will currently only work while we are in the scanLnurlWithdraw activity.

## Motivation and Context

Further extend Zap Androids capabilities.
It is a part of #159 

## How Has This Been Tested?

On my S9 using https://lnurl.bigsun.xyz/

## Screenshots (if appropriate):
receiveBSD
withdrawBSD
withdrawSuccess

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [Contribution document](../docs/CONTRIBUTING.md).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by Relaxo143** [comment]
What does it mean to withdraw in LN context? The same as receiving? So does this PR enable lnurl receiving on mobile?

**Comment by michaelWuensch** [comment]
@Relaxo143
Lets say I have an account at some service and I have 1000 sats there.
To withdraw those sats into my Zap android wallet prior to that PR I had to mannualy make an invoice for 1000 sats, send that invoice to the service and requst them to pay.
Now if the service supports it, I hit a withdraw button and a QR Code is displayed. I scan that code wit Zap Android and I am done.

**Comment by Relaxo143** [comment]
Neat!

File: app/src/main/java/zapsolutions/zap/HomeActivity.java [hunk 1]
```diff
@@ -477,4 +478,37 @@ public void onError(String error, int duration) {
             }
         });
     }
+
+    @Override
+    public void onActivityResult(int requestCode, int resultCode, Intent data) {
```
[the remark this answers is not in the document]
Comment (michaelWuensch) [reply app/src/main/java/zapsolutions/zap/HomeActivity.java:483]: Yes, I had it in the WalletFragment before, but there onActivityResult does not get called when the LNURLScanActivity is finished. It is strange as it works when directly calling the Activity from the WalletFragment, but not when it is called from the bottom sheet dialog.
I searched the internet for solutions an a lot of people said this should always be in the Activity, not the fragment. I tried it and that worked.
If you have another working solution, we can change it, but I didn't get it to work.

File: app/src/main/java/zapsolutions/zap/lnurl/ScanLnurlWithdrawActivity.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        } catch (Exception e) {
+            ZapLog.debug(LOG_TAG, e.getMessage());
+            showError(getResources().getString(R.string.lnurl_decoding_failed), 4000);
+        }
+    }
+
+    private void validateFirstResponse(@NonNull String withdrawResponse) {
+        LnurlWithdrawResponse lnurlWithdrawResponse = new Gson().fromJson(withdrawResponse, LnurlWithdrawResponse.class);
+
+        if (lnurlWithdrawResponse.getStatus() != null) {
```
[the remark this answers is not in the document]
Comment (michaelWuensch) [reply app/src/main/java/zapsolutions/zap/lnurl/ScanLnurlWithdrawActivity.java:112]: :+1:

File: app/src/main/java/zapsolutions/zap/util/LnurlDecoder.java [hunk 3]
```diff
@@ -1,17 +1,20 @@
...
 
-import zapsolutions.zap.util.Bech32;
-
 public class LnurlDecoder {
 
     public static String decode(String lnurl) throws Exception {
 
         if (lnurl == null) {
-            throw new IllegalArgumentException("Lnurl decoding failed: The data to decode is not a lnurl");
+            throw new IllegalArgumentException("LNURL decoding failed: The data to decode is not a LNURL");
```
[the remark this answers is not in the document]
Comment (michaelWuensch) [reply app/src/main/java/zapsolutions/zap/util/LnurlDecoder.java:8]: No, i don't think so.
Maybe you have noticed that these messages are actually never displayed as error messages visible to the user. These messages are to technical in my opinion and are used for debugging. 
Everything that the end user will see if the decoding fails is the already translatable message:
"The data does not contain a valid LNURL."

**Pull request merged**